File: PATTERNS/prefix_sum.py
"""
PREFIX SUM PATTERN
Use: Range sum queries, subarray sums, 2D range queries
Time: O(n) build, O(1) query, Space: O(n)
"""

from typing import List


# === 1D PREFIX SUM ===
def build_prefix_sum(arr: List[int]) -> List[int]:
    """Build prefix sum array. prefix[i] = sum(arr[0:i])"""
    n = len(arr)
    prefix = [0] * (n + 1)

    for i in range(n):
        prefix[i + 1] = prefix[i] + arr[i]

    return prefix


def range_sum(prefix: List[int], left: int, right: int) -> int:
    """Sum of arr[left:right+1] using prefix sum"""
    return prefix[right + 1] - prefix[left]


# === SUBARRAY SUM EQUALS K ===
def count_subarrays_sum_k(arr: List[int], k: int) -> int:
    """Count subarrays with sum exactly k"""
    from collections import defaultdict

    prefix_count = defaultdict(int)
    prefix_count[0] = 1

    count = 0
    prefix_sum = 0

    for num in arr:
        prefix_sum += num
        count += prefix_count[prefix_sum - k]
        prefix_count[prefix_sum] += 1

    return count


# === PRODUCT EXCEPT SELF ===
def product_except_self(arr: List[int]) -> List[int]:
    """Product of all elements except self, without division"""
    n = len(arr)
    result = [1] * n

    # left products
    left_prod = 1
    for i in range(n):
        result[i] = left_prod
        left_prod *= arr[i]

    # right products
    right_prod = 1
    for i in range(n - 1, -1, -1):
        result[i] *= right_prod
        right_prod *= arr[i]

    return result


# === 2D PREFIX SUM ===
class PrefixSum2D:
    """2D prefix sum for rectangle sum queries"""

    def __init__(self, matrix: List[List[int]]):
        if not matrix or not matrix[0]:
            self.prefix = [[]]
            return

        m, n = len(matrix), len(matrix[0])
        self.prefix = [[0] * (n + 1) for _ in range(m + 1)]

        for i in range(m):
            for j in range(n):
                self.prefix[i + 1][j + 1] = (
                    matrix[i][j]
                    + self.prefix[i][j + 1]
                    + self.prefix[i + 1][j]
                    - self.prefix[i][j]
                )

    def query(self, r1: int, c1: int, r2: int, c2: int) -> int:
        """Sum of rectangle from (r1,c1) to (r2,c2) inclusive"""
        return (
            self.prefix[r2 + 1][c2 + 1]
            - self.prefix[r1][c2 + 1]
            - self.prefix[r2 + 1][c1]
            + self.prefix[r1][c1]
        )


# === DIFFERENCE ARRAY ===
def build_difference_array(arr: List[int]) -> List[int]:
    """Build difference array for range updates"""
    n = len(arr)
    diff = [0] * n

    diff[0] = arr[0]
    for i in range(1, n):
        diff[i] = arr[i] - arr[i - 1]

    return diff


def range_add(diff: List[int], left: int, right: int, val: int) -> None:
    """Add val to arr[left:right+1] using difference array"""
    diff[left] += val
    if right + 1 < len(diff):
        diff[right + 1] -= val


def reconstruct_from_diff(diff: List[int]) -> List[int]:
    """Reconstruct original array from difference array"""
    arr = [0] * len(diff)
    arr[0] = diff[0]

    for i in range(1, len(diff)):
        arr[i] = arr[i - 1] + diff[i]

    return arr


# === RANGE UPDATE QUERIES ===
def process_range_updates(n: int, updates: List[List[int]]) -> List[int]:
    """
    Process multiple range updates efficiently.
    updates: [[left, right, val], ...]
    """
    diff = [0] * (n + 1)

    for left, right, val in updates:
        diff[left] += val
        diff[right + 1] -= val

    # prefix sum to get result
    result = [0] * n
    result[0] = diff[0]
    for i in range(1, n):
        result[i] = result[i - 1] + diff[i]

    return result


# === RUNNING SUM ===
def running_sum(arr: List[int]) -> List[int]:
    """Running sum of array"""
    result = []
    total = 0

    for num in arr:
        total += num
        result.append(total)

    return result


# === PIVOT INDEX ===
def pivot_index(arr: List[int]) -> int:
    """Find index where left sum equals right sum"""
    total = sum(arr)
    left_sum = 0

    for i, num in enumerate(arr):
        if left_sum == total - left_sum - num:
            return i
        left_sum += num

    return -1


# === SUBARRAY SUM DIVISIBLE BY K ===
def subarrays_div_by_k(arr: List[int], k: int) -> int:
    """Count subarrays with sum divisible by k"""
    from collections import defaultdict

    remainder_count = defaultdict(int)
    remainder_count[0] = 1

    count = 0
    prefix_sum = 0

    for num in arr:
        prefix_sum += num
        remainder = prefix_sum % k

        # handle negative remainders
        if remainder < 0:
            remainder += k

        count += remainder_count[remainder]
        remainder_count[remainder] += 1

    return count


# === LONGEST SUBARRAY WITH SUM K ===
def longest_subarray_sum_k(arr: List[int], k: int) -> int:
    """Find longest subarray with sum exactly k"""
    prefix_index = {0: -1}
    prefix_sum = 0
    max_len = 0

    for i, num in enumerate(arr):
        prefix_sum += num

        if prefix_sum - k in prefix_index:
            max_len = max(max_len, i - prefix_index[prefix_sum - k])

        # only store first occurrence
        if prefix_sum not in prefix_index:
            prefix_index[prefix_sum] = i

    return max_len


# === CONTIGUOUS ARRAY (equal 0s and 1s) ===
def find_max_equal_01(arr: List[int]) -> int:
    """Find longest subarray with equal 0s and 1s"""
    prefix_index = {0: -1}
    count = 0
    max_len = 0

    for i, num in enumerate(arr):
        count += 1 if num == 1 else -1

        if count in prefix_index:
            max_len = max(max_len, i - prefix_index[count])
        else:
            prefix_index[count] = i

    return max_len

File: PATTERNS/prefix_sum.cpp
/*
 * PREFIX SUM PATTERN
 * Use: Range sum queries, subarray sums, 2D range queries
 * Time: O(n) build, O(1) query, Space: O(n)
 */

#include <vector>
#include <unordered_map>

using namespace std;

// === 1D PREFIX SUM ===
vector<long long> buildPrefixSum(vector<int>& arr) {
    int n = arr.size();
    vector<long long> prefix(n + 1, 0);

    for (int i = 0; i < n; i++) {
        prefix[i + 1] = prefix[i] + arr[i];
    }

    return prefix;
}

long long rangeSum(vector<long long>& prefix, int left, int right) {
    return prefix[right + 1] - prefix[left];
}

// === SUBARRAY SUM EQUALS K ===
int countSubarraysSumK(vector<int>& arr, int k) {
    unordered_map<long long, int> prefixCount;
    prefixCount[0] = 1;

    int count = 0;
    long long prefixSum = 0;

    for (int num : arr) {
        prefixSum += num;
        count += prefixCount[prefixSum - k];
        prefixCount[prefixSum]++;
    }

    return count;
}

// === PRODUCT EXCEPT SELF ===
vector<int> productExceptSelf(vector<int>& arr) {
    int n = arr.size();
    vector<int> result(n, 1);

    // left products
    int leftProd = 1;
    for (int i = 0; i < n; i++) {
        result[i] = leftProd;
        leftProd *= arr[i];
    }

    // right products
    int rightProd = 1;
    for (int i = n - 1; i >= 0; i--) {
        result[i] *= rightProd;
        rightProd *= arr[i];
    }

    return result;
}

// === 2D PREFIX SUM ===
class PrefixSum2D {
public:
    vector<vector<long long>> prefix;

    PrefixSum2D(vector<vector<int>>& matrix) {
        if (matrix.empty() || matrix[0].empty()) return;

        int m = matrix.size(), n = matrix[0].size();
        prefix.assign(m + 1, vector<long long>(n + 1, 0));

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                prefix[i + 1][j + 1] = matrix[i][j]
                    + prefix[i][j + 1]
                    + prefix[i + 1][j]
                    - prefix[i][j];
            }
        }
    }

    long long query(int r1, int c1, int r2, int c2) {
        return prefix[r2 + 1][c2 + 1]
             - prefix[r1][c2 + 1]
             - prefix[r2 + 1][c1]
             + prefix[r1][c1];
    }
};

// === DIFFERENCE ARRAY ===
vector<int> buildDifferenceArray(vector<int>& arr) {
    int n = arr.size();
    vector<int> diff(n);

    diff[0] = arr[0];
    for (int i = 1; i < n; i++) {
        diff[i] = arr[i] - arr[i - 1];
    }

    return diff;
}

void rangeAdd(vector<int>& diff, int left, int right, int val) {
    diff[left] += val;
    if (right + 1 < diff.size()) {
        diff[right + 1] -= val;
    }
}

vector<int> reconstructFromDiff(vector<int>& diff) {
    int n = diff.size();
    vector<int> arr(n);

    arr[0] = diff[0];
    for (int i = 1; i < n; i++) {
        arr[i] = arr[i - 1] + diff[i];
    }

    return arr;
}

// === RANGE UPDATE QUERIES ===
vector<int> processRangeUpdates(int n, vector<vector<int>>& updates) {
    vector<int> diff(n + 1, 0);

    for (auto& update : updates) {
        int left = update[0], right = update[1], val = update[2];
        diff[left] += val;
        diff[right + 1] -= val;
    }

    vector<int> result(n);
    result[0] = diff[0];
    for (int i = 1; i < n; i++) {
        result[i] = result[i - 1] + diff[i];
    }

    return result;
}

// === RUNNING SUM ===
vector<int> runningSum(vector<int>& arr) {
    vector<int> result;
    int total = 0;

    for (int num : arr) {
        total += num;
        result.push_back(total);
    }

    return result;
}

// === PIVOT INDEX ===
int pivotIndex(vector<int>& arr) {
    long long total = 0;
    for (int num : arr) total += num;

    long long leftSum = 0;
    for (int i = 0; i < arr.size(); i++) {
        if (leftSum == total - leftSum - arr[i]) {
            return i;
        }
        leftSum += arr[i];
    }

    return -1;
}

// === SUBARRAY SUM DIVISIBLE BY K ===
int subarraysDivByK(vector<int>& arr, int k) {
    unordered_map<int, int> remainderCount;
    remainderCount[0] = 1;

    int count = 0;
    int prefixSum = 0;

    for (int num : arr) {
        prefixSum += num;
        int remainder = ((prefixSum % k) + k) % k;  // handle negative

        count += remainderCount[remainder];
        remainderCount[remainder]++;
    }

    return count;
}

// === LONGEST SUBARRAY WITH SUM K ===
int longestSubarraySumK(vector<int>& arr, int k) {
    unordered_map<long long, int> prefixIdx;
    prefixIdx[0] = -1;

    long long prefixSum = 0;
    int maxLen = 0;

    for (int i = 0; i < arr.size(); i++) {
        prefixSum += arr[i];

        if (prefixIdx.count(prefixSum - k)) {
            maxLen = max(maxLen, i - prefixIdx[prefixSum - k]);
        }

        if (!prefixIdx.count(prefixSum)) {
            prefixIdx[prefixSum] = i;
        }
    }

    return maxLen;
}

// === CONTIGUOUS ARRAY (equal 0s and 1s) ===
int findMaxEqual01(vector<int>& arr) {
    unordered_map<int, int> prefixIdx;
    prefixIdx[0] = -1;

    int count = 0;
    int maxLen = 0;

    for (int i = 0; i < arr.size(); i++) {
        count += (arr[i] == 1) ? 1 : -1;

        if (prefixIdx.count(count)) {
            maxLen = max(maxLen, i - prefixIdx[count]);
        } else {
            prefixIdx[count] = i;
        }
    }

    return maxLen;
}
